config/initializers/access_token_authentication.rb

```diff
@@ -1,21 +1,23 @@
 # frozen_string_literal: true
 
-ActiveSupport.on_load(:active_record) do
-  if defined?(ActiveRecord::ConnectionAdapters::Mysql2Adapter)
-    ActiveRecord::ConnectionAdapters::Mysql2Adapter.class_eval do
-      include ApiAuthentication::ByAccessToken::ConnectionExtension
+Rails.application.config.to_prepare do
```

Comment:
Why did you wrap this under `to_prepare`?

If I'm not wrong, to_prepare runs:

- once after initialization is done (production and development)
- once after every code reload (file change in development)

https://guides.rubyonrails.org/v7.0/configuring.html#initialization-events

If it runs after initialization, I assume ActiveRecord is already loaded, so no need to `ActiveSupport.on_load(:active_record)`.

This adds a callback to run when `ActiveRecord` is loaded, but shouldn't it already be loaded when calling `to_prepare` callbacks?. In that case the wrapped code would never be executed because ActiveRecord is already loaded.

On the other hand, if we asume `ActiveRecord` is reloaded every time a file changes (I'm not sure this is true), then the wrapped code would be executed multiple times, once at every code reload.

For those reasons I would remove `Rails.application.config.to_prepare`. I think it was fine before. Did you notice some problem?

Reply:
ActiveRecord might be loaded before initialization finishes. Maybe we need both - after initialization and after activerecord was loaded. But it might be worth understanding better how this works.

Reply:
Unfortunately, I lost my notes about this... But essentially - I don't quite understand the details, but without this wrapping I was seeing errors. Then I found some other project that fixed a similar issue by wrapping into `to_prepare`, and thought it was fine. That did resolve the issue.

I'll try to reproduce this again and see if I remember.

Reply:
So, you're right that this is not needed when running the server (at least in dev mode). But the `db:create` task fails without `to_prepare` wrapping.
The error is always the same, but I'll copy each error for its respective initilizer.

```
$ RAILS_ENV=test bundle exec rake ci:db:ready db:create
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
Connected to mysql2://root@127.0.0.1/systemdb0
rake aborted!
NameError: uninitialized constant ApiAuthentication

      include ApiAuthentication::ByAccessToken::ConnectionExtension
              ^^^^^^^^^^^^^^^^^
/home/dmayorov/Projects/3scale/porta/config/initializers/access_token_authentication.rb:6:in `block (2 levels) in <main>'
/home/dmayorov/Projects/3scale/porta/config/initializers/access_token_authentication.rb:5:in `class_eval'
/home/dmayorov/Projects/3scale/porta/config/initializers/access_token_authentication.rb:5:in `block in <main>'
/home/dmayorov/Projects/3scale/porta/config/initializers/access_token_authentication.rb:3:in `<main>'
/home/dmayorov/Projects/3scale/porta/config/environment.rb:8:in `<main>'
/home/dmayorov/.asdf/installs/ruby/3.1.5/bin/bundle:25:in `load'
/home/dmayorov/.asdf/installs/ruby/3.1.5/bin/bundle:25:in `<main>'
Tasks: TOP => db:create => db:load_config => environment
(See full trace by running task with --trace)
```

Reply:
Apparently `ApiAuthentication` is not loaded at the moment this initializer runs, that's why `to_prepare` fixes it, because it runs after the whole app is initialized.

Why it worked in Rails 6? I don't know, maybe Rails loads components differently in that version, I don't think it's worth trying to figure out, it must be just internal changes.

So we assume `to_prepare` is required, OK, then IMO we can assume `ActiveRecord` will be already loaded when `to_prepare` callbacks are executed. You can place a breakpoint there and verify it. If it's loaded, we can safely remove `ActiveSupport.on_load(:active_record)`. If not, we can think of a solution later, so let's assume `ActiveRecord` is loaded.

In this scenario:
1. we don't need `if defined?(ActiveRecord::ConnectionAdapters::Mysql2Adapter)`, just verify it's loaded with the debugger
2. Since the code will be executed multiple times, after every file change, we might want to implement a mechanism to not `include ApiAuthentication::ByAccessToken::ConnectionExtension` multiple times. But since this would only happen in development and not production, maybe it's not so important.

Reply:
As I suggested before, if the `on_load` block survives reload, then try `after_initialize` instead of `to_prepare`. To avoid multiple blocks being registered in dev mode. I have no other concerns about it.

Reply:
But `after_initialize` is executed before `to_prepare`, so it's possible `ActiveRecord` is not loaded at that point. But I agree, if `ActiveRecord` is already loaded when calling `after_initialize`, then that's better because it runs only once during startup